# Домашнее задание «Библиотека Pandas»

**IMPORTANT** Все дата файлы должны быть в папке data рядом с данным скриптом. Файлы данных не добавлены в git репозиторий.


In [1]:
import pandas as pd

## Задание 1
Скачайте с сайта https://grouplens.org/datasets/movielens/ датасет любого размера. Определите какому фильму было выставлено больше всего оценок 5.0.

In [6]:
!ls ./data

ml-latest-small.zip power.csv           transactions.csv


In [21]:
movies = pd.read_csv('./data/ml-latest-small/movies.csv')
ratings = pd.read_csv('./data/ml-latest-small/ratings.csv')
print('MOVIES', len(movies))
print(movies.head())
print('RATINGS', len(ratings))
print(ratings.head())

MOVIES 9742
   movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy  
RATINGS 100836
   userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931


In [61]:
best_movie_id = ratings[ ratings['rating'] == 5]['movieId'].value_counts().idxmax()
best_movie = movies[movies['movieId'] == best_movie_id]
best_movie.iloc[0]['title']

'Shawshank Redemption, The (1994)'

Фильм с наибольшим числом оценок 5 -- **The Shawshank Redemption (1994)**

## Задание 2

По данным файла `power.csv` посчитайте суммарное потребление стран Прибалтики (Латвия, Литва и Эстония) категорий 4, 12 и 21 за период с 2005 по 2010 года. Не учитывайте в расчетах отрицательные значения `quantity`.

In [63]:
!ls ./data/


ml-latest-small     ml-latest-small.zip power.csv           transactions.csv


In [66]:
power = pd.read_csv('data/power.csv')
power.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1189482 entries, 0 to 1189481
Data columns (total 4 columns):
country     1189482 non-null object
year        1189482 non-null int64
quantity    1189482 non-null float64
category    1189482 non-null int64
dtypes: float64(1), int64(2), object(1)
memory usage: 36.3+ MB


In [88]:
baltic = ['Latvia', 'Estonia', 'Lithuania']
categories = [4, 12, 21]

total_quantity = power[ (power['country'].isin(baltic))
      & (power['category'].isin(categories)) 
      & (power['year'].between(2005, 2010, True))
      & (power['quantity'] > 0)]['quantity'].sum()
print(total_quantity)

240580.0


Ответ: Суммарное потребление стран Прибалтики категорий 4, 12 и 21 за период с 2005 по 2010 года (включительно) -- **240580**

## Задание 3

Выберите страницу любого сайта с табличными данными. Импортируйте таблицы в pandas dataframe.

In [98]:
raw = pd.read_html('https://www.cbr.ru/currency_base/daily/', decimal=',', thousands = '.')
currency = raw[0]
print(currency.info())
print(currency.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 5 columns):
Цифр. код    34 non-null int64
Букв. код    34 non-null object
Единиц       34 non-null int64
Валюта       34 non-null object
Курс         34 non-null float64
dtypes: float64(1), int64(2), object(2)
memory usage: 1.4+ KB
None
   Цифр. код Букв. код  Единиц                 Валюта     Курс
0         36       AUD       1   Австралийский доллар  58.2249
1        944       AZN       1  Азербайджанский манат  43.5196
2         51       AMD     100       Армянских драмов  14.1060
3        933       BYN       1      Белорусский рубль  28.5429
4        975       BGN       1         Болгарский лев  45.8385


In [102]:
currency.sort_values('Курс', ascending=False).head(10)

Цифр. код Букв. код  Единиц                                    Валюта  \
22        960       XDR       1     СДР (специальные права заимствования)   
28        826       GBP       1  Фунт стерлингов Соединенного королевства   
8         344       HKD      10                      Гонконгских долларов   
11        978       EUR       1                                      Евро   
30        752       SEK      10                             Шведских крон   
19        578       NOK      10                           Норвежских крон   
15        417       KGS     100                          Киргизских сомов   
31        756       CHF       1                         Швейцарский франк   
10        840       USD       1                                Доллар США   
26        860       UZS   10000                           Узбекских сумов   

        Курс  
22  106.7520  
28  103.7024  
8    95.4180  
11   89.6604  
30   89.3572  
19   87.3475  
15   87.3475  
31   82.5338  
10   73.9833  
26   70.2444